Design/Workbench/design-projectsaving.md

```diff
@@ -0,0 +1,115 @@
+
+An Example X:
+```python
+def X(self):  
+    return_list = {}  
+    return_list["name"] = "TestUI"  
+    return_list["spinBox"] = self.ui.spinBox.value()  
+    return_list["lineEdit"] = self.ui.lineEdit.text()  
+    return_list["dateEdit"] = self.ui.dateEdit.text()  
+    return_list["radioButton"] = self.ui.radioButton.isChecked()      
+    return_list["checkBox"] = self.ui.checkBox.isChecked()  
+    return_list["label_7Visible"] = self.ui.label_7.isVisible()  
+    return return_list
+```
+An Example Y:
+```python
+def Y(self, dict):  
+    self.ui.spinBox.setValue(dict["spinBox"]) 
+    self.ui.lineEdit.setValue(dict["lineEdit"]) 
+    self.ui.dateEdit.setValue(dict["dateEdit"]) 
+    self.ui.radioButton.setValue(dict["radioButton"]) 
+    self.ui.checkBox.setValue(dict["checkBox"])
+    self.ui.label_7.setVisible(dict["label_7Visible"])
+```
+One problem is how the Python JSON module will represent these QStrings, QStringList and other types that are necessary to save. Inside Qt and PyQt there is a type called QVariant which allows for constructions from the majority of Qt types and normal types, which then allows you to convert them to many formats including string and QJsonArray.
+
+However the usage of QVariant is not necessary as long as you are mostly requiring QString, bool, and int, as it will just be converted to a string by the dictionary creation. Then the JSON module will be able to convert the dictionary into a JSON string.
+
+A basic example of how a line of JSON that may be generated from an object that uses the X function's output, the output is passed as the variable dict. Using indent=4 makes it a more human readable form when actually written out than with no indent as it would be on a single line:
+```python
+def json_line_from_dict(dict):  
+    import json  
+    return json.dumps(dict, indent=4) 
+ ```
+
+One potential implementation is to force usage of serialization for current and new GUIs, create a class which inherits from QWidget/QMainWindow (depending on how PyQt/Mantid Workbench implements a window), which includes a basic copy of X and Y from which windows must inherit/must implement the methods of X and Y. That was a simple list of windows can be held by the main application and a call to X and Y made on them, for saving and loading respectively. This implementation will allow for not only other methods to be forced upon GUI designers that may be required but will also guarantee there is a method for project saving.
+
+For the implementation of X and Y, we need to have a copy of the variables that needs to be serialized on the GUI, so for example if a value needs to be serialized then save it just grab the value and load it back in, in the methods, X and Y.
```

Comment:
I think you're saying that each component should only receive the state it requires to recreate itself? I agree with this.
I don't quite understand the part about saving and reloading though...

There is also a question of how we map from sections of the file, presumable string names of widgets, to actual classes and making sure this is robust over time. We might need to define some type of version...